### ISSUE TITLE
Support for neo4j 4.0

### ISSUE THREAD

**Issue by laurentS** [neo4j-contrib/neomodel#485]
I unexpectedly upgraded to neo4j 4.0 when it was pushed to debian's repositories without warning, nor supporting docs anywhere on neo4j.com.
The upgrade does not seem obvious, and it created a bunch of issues with my `neomodel` code. So I'm opening this ticket for discussion.

**Comment by aanastasiou** [comment]
Great, how about mentioning some of those issues here (?)

**Comment by KevinvOosterhout** [comment]
I do have connection problems. I use about the same configuration as i used to do with Neo4j 3.5.

```
>>> from neomodel import (config, StructuredNode, StringProperty, IntegerProperty,
...     UniqueIdProperty, RelationshipTo)
... 
... config.DATABASE_URL = 'bolt://app:ae6LuykTEN3Fa9hzwjemoULPKWfhgapu@10.211.55.37:7687'
... 
... class Country(StructuredNode):
...     code = StringProperty(unique_index=True, required=True)
... 
... class Person(StructuredNode):
...     uid = UniqueIdProperty()
...     name = StringProperty(unique_index=True)
...     age = IntegerProperty(index=True, default=0)
... 
...     # traverse outgoing IS_FROM relations, inflate to Country objects
...     country = RelationshipTo(Country, 'IS_FROM')
...     
>>> Country.nodes.all()

Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neomodel/match.py", line 472, in all
    return self.query_cls(self).build_ast()._execute(lazy)
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neomodel/match.py", line 447, in _execute
    results, _ = db.cypher_query(query, self._query_params, resolve_objects=True)
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neomodel/util.py", line 32, in wrapper
    _db.set_connection(config.DATABASE_URL)
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neomodel/util.py", line 96, in set_connection
    max_pool_size=config.MAX_POOL_SIZE)
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neo4j/__init__.py", line 120, in driver
    return Driver(uri, **config)
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neo4j/__init__.py", line 161, in __new__
    return subclass(uri, **config)
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neo4j/__init__.py", line 235, in __new__
    pool.release(pool.acquire())
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neobolt/direct.py", line 719, in acquire
    return self.acquire_direct(self.address)
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neobolt/direct.py", line 612, in acquire_direct
    connection = self.connector(address, error_handler=self.connection_error_handler)
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neo4j/__init__.py", line 232, in connector
    return connect(address, **dict(config, **kwargs))
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neobolt/direct.py", line 976, in connect
    raise last_error
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neobolt/direct.py", line 967, in connect
    s, der_encoded_server_certificate = _secure(s, host, security_plan.ssl_context, **config)
  File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neobolt/direct.py", line 858, in _secure
    s = ssl_context.wrap_socket(s, server_hostname=host if HAS_SNI and host else None)
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 423, in wrap_socket
    session=session
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 870, in _create
    self.do_handshake()
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 1139, in do_handshake
    self._sslobj.do_handshake()
OSError: [Errno 0] Error

```

**Comment by KevinvOosterhout** [comment]
> I do have connection problems. I use about the same configuration as i used to do with Neo4j 3.5.
> 
> ```
> >>> from neomodel import (config, StructuredNode, StringProperty, IntegerProperty,
> ...     UniqueIdProperty, RelationshipTo)
> ... 
> ... config.DATABASE_URL = 'bolt://app:password@10.211.55.37:7687'
> ... 
> ... class Country(StructuredNode):
> ...     code = StringProperty(unique_index=True, required=True)
> ... 
> ... class Person(StructuredNode):
> ...     uid = UniqueIdProperty()
> ...     name = StringProperty(unique_index=True)
> ...     age = IntegerProperty(index=True, default=0)
> ... 
> ...     # traverse outgoing IS_FROM relations, inflate to Country objects
> ...     country = RelationshipTo(Country, 'IS_FROM')
> ...     
> >>> Country.nodes.all()
> 
> Traceback (most recent call last):
>   File "<input>", line 1, in <module>
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neomodel/match.py", line 472, in all
>     return self.query_cls(self).build_ast()._execute(lazy)
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neomodel/match.py", line 447, in _execute
>     results, _ = db.cypher_query(query, self._query_params, resolve_objects=True)
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neomodel/util.py", line 32, in wrapper
>     _db.set_connection(config.DATABASE_URL)
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neomodel/util.py", line 96, in set_connection
>     max_pool_size=config.MAX_POOL_SIZE)
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neo4j/__init__.py", line 120, in driver
>     return Driver(uri, **config)
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neo4j/__init__.py", line 161, in __new__
>     return subclass(uri, **config)
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neo4j/__init__.py", line 235, in __new__
>     pool.release(pool.acquire())
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neobolt/direct.py", line 719, in acquire
>     return self.acquire_direct(self.address)
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neobolt/direct.py", line 612, in acquire_direct
>     connection = self.connector(address, error_handler=self.connection_error_handler)
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neo4j/__init__.py", line 232, in connector
>     return connect(address, **dict(config, **kwargs))
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neobolt/direct.py", line 976, in connect
>     raise last_error
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neobolt/direct.py", line 967, in connect
>     s, der_encoded_server_certificate = _secure(s, host, security_plan.ssl_context, **config)
>   File "/Users/kevin.oosterhout/Pythonprojects/Cloud Unchained/Api/utilities/venv/lib/python3.7/site-packages/neobolt/direct.py", line 858, in _secure
>     s = ssl_context.wrap_socket(s, server_hostname=host if HAS_SNI and host else None)
>   File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 423, in wrap_socket
>     session=session
>   File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 870, in _create
>     self.do_handshake()
>   File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/ssl.py", line 1139, in do_handshake
>     self._sslobj.do_handshake()
> OSError: [Errno 0] Error
> ```

The problem is related to the encrypted connections.  If i set the encrypted connection to false the problem is solved. 

```config.ENCRYPTED_CONNECTION = False```

**Comment by laurentS** [comment]
After upgrading `neo4j` to 4.0, I'm getting the same problem with SSH as @KevinvOosterhout described above. If I disable encryption, I move on to `neobolt.exceptions.TransientError: Database 'graph.db' is unavailable` which is due to the database format having changed since 3.5. The upgrade procedure is explained [here](https://neo4j.com/docs/migration-guide/4.0/).

Beyond this, my test suite is all broken with `CypherSyntaxError`s. This seems essentially due to the [long list of changes](https://neo4j.com/docs/cypher-manual/4.0/deprecations-additions-removals-compatibility/#cypher-deprecations-additions-removals) in CYPHER syntax. My impression is that a number of queries will have to be rewritten in neomodel.

**Comment by illmoded** [comment]
I have started clear instance of neo4j 4.0 and I'm getting:
```python
---------------------------------------------------------------------------
CypherSyntaxError                         Traceback (most recent call last)
<ipython-input-45-cefc0ff36155> in <module>
      2     d = dict(e._asdict())
      3     if d:
----> 4         company = Company.nodes.get(name="ok")
      5         print(company)

~/.local/share/virtualenvs/graphdb_test-AeDbk77K/lib/python3.7/site-packages/neomodel/match.py in get(self, lazy, **kwargs)
    549         :return: node
    550         """
--> 551         result = self._get(limit=2, lazy=lazy, **kwargs)
    552         if len(result) > 1:
    553             raise MultipleNodesReturned(repr(kwargs))

~/.local/share/virtualenvs/graphdb_test-AeDbk77K/lib/python3.7/site-packages/neomodel/match.py in _get(self, limit, lazy, **kwargs)
    540         if limit:
    541             self.limit = limit
--> 542         return self.query_cls(self).build_ast()._execute(lazy)
    543 
    544     def get(self, lazy=False, **kwargs):

~/.local/share/virtualenvs/graphdb_test-AeDbk77K/lib/python3.7/site-packages/neomodel/match.py in _execute(self, lazy)
    445             self._ast['return'] = 'id({})'.format(self._ast['return'])
    446         query = self.build_query()
--> 447         results, _ = db.cypher_query(query, self._query_params, resolve_objects=True)
    448         # The following is not as elegant as it could be but had to be copied from the
    449         # version prior to cypher_query with the resolve_objects capability.

~/.local/share/virtualenvs/graphdb_test-AeDbk77K/lib/python3.7/site-packages/neomodel/util.py in wrapper(self, *args, **kwargs)
     31         if not _db.url:
     32             _db.set_connection(config.DATABASE_URL)
---> 33         return func(self, *args, **kwargs)
     34 
     35     return wrapper

~/.local/share/virtualenvs/graphdb_test-AeDbk77K/lib/python3.7/site-packages/neomodel/util.py in cypher_query(self, query, params, handle_unique, retry_on_session_expire, resolve_objects)
    227                 exc_info = sys.exc_info()
    228                 if sys.version_info >= (3, 0):
--> 229                     raise exc_info[1].with_traceback(exc_info[2])
    230                 else:
    231                     raise exc_info[1]

~/.local/share/virtualenvs/graphdb_test-AeDbk77K/lib/python3.7/site-packages/neomodel/util.py in cypher_query(self, query, params, handle_unique, retry_on_session_expire, resolve_objects)
    210             # Retrieve the data
    211             start = time.time()
--> 212             response = session.run(query, params)
    213             results, meta = [list(r.values()) for r in response], response.keys()
    214             end = time.time()

~/.local/share/virtualenvs/graphdb_test-AeDbk77K/lib/python3.7/site-packages/neo4j/__init__.py in run(self, statement, parameters, **kwparameters)
    490             try:
    491                 self._connection.send()
--> 492                 self._connection.fetch()
    493             except ConnectionExpired as error:
    494                 raise SessionExpired(*error.args)

~/.local/share/virtualenvs/graphdb_test-AeDbk77K/lib/python3.7/site-packages/neobolt/direct.py in fetch(self)
    417     def fetch(self):
    418         try:
--> 419             return self._fetch()
    420         except self.error_handler.known_errors as error:
    421             self.error_handler.handle(error, self.unresolved_address)

~/.local/share/virtualenvs/graphdb_test-AeDbk77K/lib/python3.7/site-packages/neobolt/direct.py in _fetch(self)
    459         elif summary_signature == b"\x7F":
    460             log_debug("[#%04X]  S: FAILURE %r", self.local_port, summary_metadata)
--> 461             response.on_failure(summary_metadata or {})
    462         else:
    463             raise ProtocolError("Unexpected response message with signature %02X" % summary_signature)

~/.local/share/virtualenvs/graphdb_test-AeDbk77K/lib/python3.7/site-packages/neobolt/direct.py in on_failure(self, metadata)
    753         if callable(handler):
    754             handler()
--> 755         raise CypherError.hydrate(**metadata)
    756 
    757     def on_ignored(self, metadata=None):

CypherSyntaxError: The old parameter syntax `{param}` is no longer supported. Please use `$param` instead (line 1, column 49 (offset: 48))
" MATCH ((company:Company)) WHERE company.name = {company_name_1} RETURN company LIMIT 2"
```
when I do 
```python
company = Company.nodes.get(name='ok')
```
with model:
```python
class Company(StructuredNode):
    uid = UniqueIdProperty()
    name = StringProperty(default=None)
    founded_date = RelationshipTo(Date, 'FOUNDED_ON')
    closed_date = RelationshipTo(Date, 'CLOSED_ON')
    address = RelationshipTo(Address, 'HAS')
```

**Comment by KevinvOosterhout** [comment]
> After upgrading `neo4j` to 4.0, I'm getting the same problem with SSH as @KevinvOosterhout described above. If I disable encryption, I move on to `neobolt.exceptions.TransientError: Database 'graph.db' is unavailable` which is due to the database format having changed since 3.5. The upgrade procedure is explained [here](https://neo4j.com/docs/migration-guide/4.0/).
> 
> Beyond this, my test suite is all broken with `CypherSyntaxError`s. This seems essentially due to the [long list of changes](https://neo4j.com/docs/cypher-manual/4.0/deprecations-additions-removals-compatibility/#cypher-deprecations-additions-removals) in CYPHER syntax. My impression is that a number of queries will have to be rewritten in neomodel.
> -- EDIT --
> I tried upgrading from neo4j 3.5.x. I use the community edition, if it makes any difference.

Could you try adding the `dbms.default_database` setting in the neo4j config? The value should be the database name you're using -'.db' so for graph.db it would be graph. Could imagine the new multi-database support is having some issue and this is selecting the default database you want to use by default.

**Comment by KevinvOosterhout** [comment]
> ```python
> CypherSyntaxError: The old parameter syntax `{param}` is no longer supported. Please use `$param` instead
> ```

I would recommend creating a new issue for this. Neo4j dropped {} variables for the $ ones. This was already deprecated in 3.5. It's probably time to replace all of those and drop support for releases that don't support it or make a setting to choose which ones and let the new default be the $ variables

**Comment by laurentS** [comment]
> > After upgrading `neo4j` to 4.0, I'm getting the same problem with SSH as @KevinvOosterhout described above. If I disable encryption, I move on to `neobolt.exceptions.TransientError: Database 'graph.db' is unavailable` which is due to the database format having changed since 3.5. The upgrade procedure is explained [here](https://neo4j.com/docs/migration-guide/4.0/).
> 
> Could you try adding the `dbms.default_database` setting in the neo4j config? The value should be the database name you're using -'.db' so for graph.db it would be graph. Could imagine the new multi-database support is having some issue and this is selecting the default database you want to use by default.

I forgot to mention that in the neo4j logs, I had a message telling me to set `dbms.allow_upgrade=true` in the config. The setting allowed the server to convert my `graph.db` to the 4.0 format (which is not a reversible action). It's also described in the [migration docs](https://neo4j.com/docs/migration-guide/4.0/upgrade-single-instance/).

**Comment by foxban** [comment]
I got the following Exception after upgrading neo4j server to 4.0.0

neomodel version is 3.3.2

```
Traceback (most recent call last):
  File "txs_neo4j.py", line 61, in <module>
    n_init_addr = Address.nodes.get_or_none(addr=init_addr)
  File "/home/wangli7/.local/lib/python2.7/site-packages/neomodel/match.py", line 567, in get_or_none
    return self.get(**kwargs)
  File "/usr/local/lib/python2.7/site-packages/neomodel/match.py", line 551, in get
    result = self._get(limit=2, lazy=lazy, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/neomodel/match.py", line 542, in _get
    return self.query_cls(self).build_ast()._execute(lazy)
  File "/usr/local/lib/python2.7/site-packages/neomodel/match.py", line 447, in _execute
    results, _ = db.cypher_query(query, self._query_params, resolve_objects=True)
  File "/usr/local/lib/python2.7/site-packages/neomodel/util.py", line 33, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/neomodel/util.py", line 231, in cypher_query
    raise exc_info[1]
neobolt.exceptions.CypherSyntaxError: The old parameter syntax `{param}` is no longer supported. Please use `$param` instead (line 1, column 49 (offset: 48))
" MATCH ((address:Address)) WHERE address.addr = {address_addr_1} RETURN address LIMIT 2"
                                                 ^
```

**Comment by aanastasiou** [comment]
Thanks, some of these notes have been very useful.

**Comment by CoreTex** [comment]
Could someone please merge this to fix the issue.

**Comment by jcrogel** [comment]
After installing SSL certificates and getting past the SSL issues I get this

```
---------------------------------------------------------------------------
CypherSyntaxError                         Traceback (most recent call last)
<ipython-input-19-2aa31ae648e0> in <module>
----> 1 case = NeoCase.nodes.get_or_none(case_uid="XXXXXXXX")

/usr/local/lib/python3.7/site-packages/neomodel/match.py in get_or_none(self, **kwargs)
    565         """
    566         try:
--> 567             return self.get(**kwargs)
    568         except self.source_class.DoesNotExist:
    569             pass

/usr/local/lib/python3.7/site-packages/neomodel/match.py in get(self, lazy, **kwargs)
    549         :return: node
    550         """
--> 551         result = self._get(limit=2, lazy=lazy, **kwargs)
    552         if len(result) > 1:
    553             raise MultipleNodesReturned(repr(kwargs))

/usr/local/lib/python3.7/site-packages/neomodel/match.py in _get(self, limit, lazy, **kwargs)
    540         if limit:
    541             self.limit = limit
--> 542         return self.query_cls(self).build_ast()._execute(lazy)
    543 
    544     def get(self, lazy=False, **kwargs):

/usr/local/lib/python3.7/site-packages/neomodel/match.py in _execute(self, lazy)
    445             self._ast['return'] = 'id({})'.format(self._ast['return'])
    446         query = self.build_query()
--> 447         results, _ = db.cypher_query(query, self._query_params, resolve_objects=True)
    448         # The following is not as elegant as it could be but had to be copied from the
    449         # version prior to cypher_query with the resolve_objects capability.

/usr/local/lib/python3.7/site-packages/neomodel/util.py in wrapper(self, *args, **kwargs)
     31             _db.set_connection(config.DATABASE_URL)
     32         return func(self, *args, **kwargs)
---> 33 
     34     return wrapper
     35 

/usr/local/lib/python3.7/site-packages/neomodel/util.py in cypher_query(self, query, params, handle_unique, retry_on_session_expire, resolve_objects)
    227                 if sys.version_info >= (3, 0):
    228                     raise exc_info[1].with_traceback(exc_info[2])
--> 229                 else:
    230                     raise exc_info[1]
    231         except SessionError:

/usr/local/lib/python3.7/site-packages/neomodel/util.py in cypher_query(self, query, params, handle_unique, retry_on_session_expire, resolve_objects)
    210             start = time.time()
    211             response = session.run(query, params)
--> 212             results, meta = [list(r.values()) for r in response], response.keys()
    213             end = time.time()
    214 

/usr/local/lib/python3.7/site-packages/neo4j/__init__.py in run(self, statement, parameters, **kwparameters)
    497             on_summary=lambda: result.detach(sync=False),
    498         )
--> 499 
    500         if not has_transaction:
    501             try:

/usr/local/lib/python3.7/site-packages/neobolt/direct.py in fetch(self)
    420     def fetch(self):
    421         try:
--> 422             return self._fetch()
    423         except self.error_handler.known_errors as error:
    424             self.error_handler.handle(error, self.unresolved_address)

/usr/local/lib/python3.7/site-packages/neobolt/direct.py in _fetch(self)
    462         elif summary_signature == b"\x7F":
    463             log_debug("[#%04X]  S: FAILURE %r", self.local_port, summary_metadata)
--> 464             response.on_failure(summary_metadata or {})
    465         else:
    466             raise ProtocolError("Unexpected response message with signature %02X" % summary_signature)

/usr/local/lib/python3.7/site-packages/neobolt/direct.py in on_failure(self, metadata)
    757         if callable(handler):
    758             handler()
--> 759         raise CypherError.hydrate(**metadata)
    760 
    761     def on_ignored(self, metadata=None):

CypherSyntaxError: The old parameter syntax `{param}` is no longer supported. Please use `$param` instead (line 1, column 44 (offset: 43))
" MATCH ((case:Case)) WHERE case.case_uid = {case_case_uid_1} RETURN case LIMIT 2"
```

**Comment by kamTheNigerian** [comment]
> After installing SSL certificates and getting past the SSL issues I get this
> 
> ```
> ---------------------------------------------------------------------------
> CypherSyntaxError                         Traceback (most recent call last)
> <ipython-input-19-2aa31ae648e0> in <module>
> ----> 1 case = NeoCase.nodes.get_or_none(case_uid="XXXXXXXX")
> 
> /usr/local/lib/python3.7/site-packages/neomodel/match.py in get_or_none(self, **kwargs)
>     565         """
>     566         try:
> --> 567             return self.get(**kwargs)
>     568         except self.source_class.DoesNotExist:
>     569             pass
> 
> /usr/local/lib/python3.7/site-packages/neomodel/match.py in get(self, lazy, **kwargs)
>     549         :return: node
>     550         """
> --> 551         result = self._get(limit=2, lazy=lazy, **kwargs)
>     552         if len(result) > 1:
>     553             raise MultipleNodesReturned(repr(kwargs))
> 
> /usr/local/lib/python3.7/site-packages/neomodel/match.py in _get(self, limit, lazy, **kwargs)
>     540         if limit:
>     541             self.limit = limit
> --> 542         return self.query_cls(self).build_ast()._execute(lazy)
>     543 
>     544     def get(self, lazy=False, **kwargs):
> 
> /usr/local/lib/python3.7/site-packages/neomodel/match.py in _execute(self, lazy)
>     445             self._ast['return'] = 'id({})'.format(self._ast['return'])
>     446         query = self.build_query()
> --> 447         results, _ = db.cypher_query(query, self._query_params, resolve_objects=True)
>     448         # The following is not as elegant as it could be but had to be copied from the
>     449         # version prior to cypher_query with the resolve_objects capability.
> 
> /usr/local/lib/python3.7/site-packages/neomodel/util.py in wrapper(self, *args, **kwargs)
>      31             _db.set_connection(config.DATABASE_URL)
>      32         return func(self, *args, **kwargs)
> ---> 33 
>      34     return wrapper
>      35 
> 
> /usr/local/lib/python3.7/site-packages/neomodel/util.py in cypher_query(self, query, params, handle_unique, retry_on_session_expire, resolve_objects)
>     227                 if sys.version_info >= (3, 0):
>     228                     raise exc_info[1].with_traceback(exc_info[2])
> --> 229                 else:
>     230                     raise exc_info[1]
>     231         except SessionError:
> 
> /usr/local/lib/python3.7/site-packages/neomodel/util.py in cypher_query(self, query, params, handle_unique, retry_on_session_expire, resolve_objects)
>     210             start = time.time()
>     211             response = session.run(query, params)
> --> 212             results, meta = [list(r.values()) for r in response], response.keys()
>     213             end = time.time()
>     214 
> 
> /usr/local/lib/python3.7/site-packages/neo4j/__init__.py in run(self, statement, parameters, **kwparameters)
>     497             on_summary=lambda: result.detach(sync=False),
>     498         )
> --> 499 
>     500         if not has_transaction:
>     501             try:
> 
> /usr/local/lib/python3.7/site-packages/neobolt/direct.py in fetch(self)
>     420     def fetch(self):
>     421         try:
> --> 422             return self._fetch()
>     423         except self.error_handler.known_errors as error:
>     424             self.error_handler.handle(error, self.unresolved_address)
> 
> /usr/local/lib/python3.7/site-packages/neobolt/direct.py in _fetch(self)
>     462         elif summary_signature == b"\x7F":
>     463             log_debug("[#%04X]  S: FAILURE %r", self.local_port, summary_metadata)
> --> 464             response.on_failure(summary_metadata or {})
>     465         else:
>     466             raise ProtocolError("Unexpected response message with signature %02X" % summary_signature)
> 
> /usr/local/lib/python3.7/site-packages/neobolt/direct.py in on_failure(self, metadata)
>     757         if callable(handler):
>     758             handler()
> --> 759         raise CypherError.hydrate(**metadata)
>     760 
>     761     def on_ignored(self, metadata=None):
> 
> CypherSyntaxError: The old parameter syntax `{param}` is no longer supported. Please use `$param` instead (line 1, column 44 (offset: 43))
> " MATCH ((case:Case)) WHERE case.case_uid = {case_case_uid_1} RETURN case LIMIT 2"
> ```

I am running into this same error has there been a fix by chance

**Comment by kamTheNigerian** [comment]
hey guys.. I am running into issues with neomodel v3.3.2 and neo4j v1.7.6
neobolt.exceptions.CypherSyntaxError: The old parameter syntax {param} is no longer supported. Please use $param instead (line 1, column 18 (offset: 17))
"CREATE (n:Person {create_params}) RETURN n"


any help its been driving me nuts

**Comment by KevinvOosterhout** [comment]
@jcrogel @kamTheNigerian This is due to the main branch using the old parameter style. You need to run Neo4j 4.x in Cypher 3.5 mode. You can change this in your neo4j.conf  by setting the option: `cypher.default_language_version=3.5`.

**Comment by aanastasiou** [comment]
Can you please update to version 4.0.0 (of neomodel) and let us know if this still persists?

**Comment by gitkp11** [comment]
**Still the same issue persists in latest version till today's date. I have created issue for it in neomodel. Please can someone check it.**


**CypherSyntaxError:The old parameter syntax `{param}` is no longer supported. Please use `$param` instead. in noemodel v3.3 not in neo4j v4.1**

This issue is resolved in neo4j v4.1 update. But my question is related to neomodel v3.3. How do we resolve this issue in neomodel v3.3.

My queries: I have models.py as
```
from neomodel import StructuredNode, StringProperty, UniqueIdProperty()

class Person(StructuredNode):
  person_id = UniqueIdProperty()
  person_name = StringProperty()
```
when i execute
`python manage.py install_labels` it works fine with no errors
```
$ python manage.py install_labels
Setting up indexes and constraints...

Found django_neomodel.DjangoNode
 ! Skipping class django_neomodel.DjangoNode is abstract
Found neo4japp.models.Person
 + Creating unique constraint for person_id on label Person for class neo4japp.models.Person

Finished 2 classes.
```
But when I create its object like

`modiji = Person(person_id='1', person_name='modiji').save()`

it gives me this error
```
CypherSyntaxError: The old parameter syntax `{param}` is no longer supported. Please use `$param` instead (line 1, column 18 (offset: 17)) "CREATE (n:Person {create_params}) RETURN n"
```
As in OGM we cant manipulate cypher queries how do i make it work. Please help. I want to make use of OGM and dont want to hard code cypher queries to make my code effectively using django framework & neomodel

**If not then what is the alternative to neomodel to use OGM queries like the one in ORM so as to ease the database quering stuff?**
